src/main/java/frc/robot/subsystems/SwerveModule.java

```diff
@@ -0,0 +1,147 @@
+
+        this.ID = ID;
+
+        absoluteEncoder = Robot.isSimulation() ? null : AbsoluteEncoder.constructEncoder(absoluteEncoderConfig);
+
+        driveMotor = MotorController.constructMotor(driveMotorConfig);
+        turningMotor = MotorController.constructMotor(turningMotorConfig);
+
+        driveEncoder = this.driveMotor.getEncoder();
+        turningEncoder = this.turningMotor.getEncoder();
+
+        driveEncoder.setPositionConversionFactor(SwerveModuleConstants.kDriveEncoderRotFactor);
+        driveEncoder.setVelocityConversionFactor(SwerveModuleConstants.kDriveEncoderRPMFactor);
+        turningEncoder.setPositionConversionFactor(SwerveModuleConstants.kTurningEncoderRotFactor);
+        turningEncoder.setVelocityConversionFactor(SwerveModuleConstants.kTurningEncoderRPMFactor);
+
+        simDriveMotor = new FlywheelSim(
+            LinearSystemId.identifyVelocitySystem(2, 1.24), //TODO: Update with real SysID
+            DCMotor.getNEO(1),
+            SwerveModuleConstants.kDriveMotorGearRatio
+        );
+        simTurningMotor = new FlywheelSim(
+            LinearSystemId.identifyVelocitySystem(0.16, 0.0348), //TODO: Update with real SysID
+            DCMotor.getNEO(1),
+            SwerveModuleConstants.kTurningMotorGearRatio
+        );
+
+        simDriveEncoder = new RelativeEncoderSim(driveMotor);
+        simTurningEncoder = new RelativeEncoderSim(turningMotor);
+        
+        turningPIDController = new PIDController(SwerveModuleConstants.kPTurning, 0, 0);
+        turningPIDController.enableContinuousInput(-Math.PI, Math.PI);
+
+        resetEncoders();
+    }
+
+    public double getDrivePosition() {
+        return driveEncoder.getPosition();
+    }
+
+    public double getTurningPosition() {
+        return turningEncoder.getPosition();
+    }
+
+    public double getDriveVelocity() {
+        return Robot.isSimulation() ? simDriveEncoder.getVelocity() : driveEncoder.getVelocity();
+    }
+    
+
+    public double getTurningVelocity() {
+        return Robot.isSimulation() ? simTurningEncoder.getVelocity() : turningEncoder.getVelocity();
+    }
+
+    public void resetEncoders() {
+        driveEncoder.setPosition(0);
+        turningEncoder.setPosition(Robot.isSimulation() ? 0 : Units.degreesToRotations(absoluteEncoder.getAbsolutePosition()));
+    }
+
+    public SwerveModuleState getState() {
+        return new SwerveModuleState(getDriveVelocity(), new Rotation2d(getTurningPosition()));
+    }
+
+    public void setDesiredState(SwerveModuleState state) {
+        if (Math.abs(state.speedMetersPerSecond) < 0.001) { //Prevent wheels to returning to heading of 0 when controls released
+            stop();
+            return;
+        }
+        state = SwerveModuleState.optimize(state, getState().angle);
+        driveMotor.set(state.speedMetersPerSecond / DriveConstants.kPhysicalMaxSpeed);
```

Comment:
This is also a problem. You have a value in meters per second and divide it by kPhysicalMaxSpeed(1.0) and pass it to the motor controller.set method which is (-1, 1) value bound.

Reply:
Is this a problem?
In convertToModuleStates, the x, y, and r inputs, which have a range of [-1,1], are multiplied by the maximum speed to convert them to meters per second.
These are then passed on to construct chassisSpeeds, which adjusts the speeds for rotation by making them smaller based on the desired omega
Theses chassis speeds are then used by normalizeDrive (Which may have its own issues), but the speeds cannot possibly return as larger than they were before normalization due to the Math.min
These speeds are then converted back to a number in the range [-1,1] by using the same conversion factor

So in this case, the m/s speeds are being converted back to motor inputs using 1, because they were converted to m/s using 1

Reply:
I think there is a misunderstanding of how to normalize a value. The division should be by the maximum of the same unit. So kPhysicalMaxSpeed should be the absolute value of the max speed in m/s if you want to use it to normalize a m/s value. 
e.g. 10 m/s / 1.0 = 10 m/s, not mapped to -1 to 1, but 10 m/s / 50 m/s = 0.2, a normalized value.

Math.min in normalizeDrive appears like it will only saturate speeds rather than normalize them (i.e. even small input speeds will hit the max motor value. It won't exceed 1, but it also won't go below it often.)

Reply:
I'm trying to work through this
So joystick inputs are in a range of [-1,1], I suppose we'll call these units %
To convert % to m/s, we multiply this input by a unit factor, kPhysicalMaxSpeed, 1.0m/s
That speed, now in m/s, is converted to module states that use m/s
These states are used by normalizeDrive(), which multiplies these states by a scale that cannot be more than 1. normalizeDrive() is still in m/s
Then, we divide by our unit factor,  kPhysicalMaxSpeed, to convert m/s back to % to be used by the motor.

Since the initial joystick value cannot be larger than one, state.speedMetersPerSecond cannot be larger than kPhysicalMaxSpeed, so the input supplied to the motor will always be less than or equal to one.

Reply:
I misread how the normalizeDrive method was working. I still have some concerns about it but I'll address them separately.